django/db/models/aggregates.py

```diff
@@ -174,6 +175,22 @@ def __init__(self, expression, filter=None, **extra):
             raise ValueError("Star cannot be used with filter. Please specify a field.")
         super().__init__(expression, filter=filter, **extra)
 
+    def resolve_expression(self, *args, **kwargs):
+        result = super().resolve_expression(*args, **kwargs)
+        expr = result.source_expressions[0]
+
+        # In case of composite primary keys, count the first column.
+        if isinstance(expr, ColPairs):
+            if self.distinct:
+                raise NotSupportedError(
+                    "COUNT(DISTINCT) doesn't support composite primary keys"
+                )
+
+            cols = expr.get_cols()
+            return Count(cols[0], filter=result.filter)
```

Comment:
Just a small note that if we add support for composite fields in the future and they rely on `ColPairs` as well this code will break as there might `NULL` in there.

This currently only works because all cols of a `CompositePrimaryKey` which has references to it turned into `ColPairs` must be `NOT NULL` so counting one member of the composite is equivalent to counting all of them.

On non-primary key composite fields this would require doing

```python
cols = expr.get_cols()
cols_filter = result.filter
for col in cols[1:]:
   cols_filter &= Q(IsNull(col, False))
return Count(cols[0], filter=composite_filter)
```

Reply:
Yes, I think for composite PKs we should count one field, and for other composite fields in the future do it the way you're proposing. 👍